### PR TITLE
[WIP / Discussion] Add guzzle for proper proxy and SSL support

### LINKED ISSUE #9054: Improve HTTPS cert validation for appstore
make code use https interface of appstore and properly validate it's certificate

cc @karlitschek

### PR DESCRIPTION
**Pull request by LukasReschke**
This adds a wrapper around Guzzle to allow developers to perform HTTP requests using a somewhat saner API. **Without this SSL and proxy support is basically broken in ownCloud as stated in https://github.com/owncloud/core/pull/14913#issuecomment-82490848**
## Todo
- [x] Get rid of private API
- [x] Deprecate HTTPHelper
- [x] Implement SSL in such a way that it actually works……
- [x] Follow proxy configured in config.php
- [x] Determine how to replace a users' custom trusted root certificate list
- [x] Fix unit tests after we agreed on an API design
- [x] Ensure that S2S still works properly

Requires https://github.com/owncloud/3rdparty/pull/167

Fixes https://github.com/owncloud/core/issues/9054, https://github.com/owncloud/core/issues/14963, https://github.com/owncloud/core/issues/14840 and a lot of other issues

cc @bantu @nickvergessen @DeepDiver1975

### CODE REVIEW

**Comment by LukasReschke** [comment]
Especially a review from an API PoV would be welcome – any feedback welcome :smile:

File: lib/private/ocsclient.php [hunk 1]
```diff
@@ -114,14 +122,21 @@ public static function getApplications($categories, $page, $filter) {
...
+
+		$client = \OC::$server->getHTTPClientService()->newClient();
+		try {
+			$response = $client->get($url, ['timeout' => 5]);
+		} catch(\Exception $e) {
+			return null;
+		}
 
-		if ($xml == false) {
+		if($response->getStatusCode() !== (int)200) {
```
Comment (nickvergessen) [inline lib/private/ocsclient.php]: (int)200 ? :ghost:

File: lib/public/http/client/iclient.php [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+ */
+
+namespace OCP\Http\Client;
+
+/**
+ * Interface IClient
+ *
+ * @package OCP\Http
+ */
+interface IClient {
```
Comment (nickvergessen) [inline lib/public/http/client/iclient.php]: Guzzle has no interface we could extend here?

File: lib/private/ocsclient.php (hunk 1 above)
Comment (LukasReschke) [inline lib/private/ocsclient.php]: Right. Let me cast the status code at our wrapper level  :see_no_evil:

File: lib/public/http/client/iresponse.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+namespace OCP\Http\Client;
+
+use Guzzle\Http\Message\Response as GuzzleResponse;
+
+/**
+ * Interface IResponse
+ *
+ * @package OC\Http
+ */
+interface IResponse {
```
Comment (nickvergessen) [inline lib/public/http/client/iresponse.php]: same here?

File: lib/public/http/client/iclient.php (hunk 2 above)
Comment (LukasReschke) [inline lib/public/http/client/iclient.php]: We could extend `\Guzzle\Http\ClientInterface` here - does that make sense?

File: lib/public/http/client/iresponse.php (hunk 3 above)
Comment (LukasReschke) [inline lib/public/http/client/iresponse.php]: `\GuzzleHttp\Message\ResponseInterface`?

File: lib/public/http/client/iclient.php (hunk 2 above)
Comment (nickvergessen) [inline lib/public/http/client/iclient.php]: I'd say it does

File: lib/private/http/client/streaminterface.php [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ */
+
+namespace OC\Http\Client;
+
+/**
+ * Interface StreamInterface
+ *
+ * @package OC\Http
+ */
+interface StreamInterface extends \GuzzleHttp\Stream\StreamInterface {}
```
Comment (DeepDiver1975) [inline lib/private/http/client/streaminterface.php]: what for we need this?

File: lib/public/http/client/iclient.php (hunk 2 above)
Comment (DeepDiver1975) [inline lib/public/http/client/iclient.php]: no 3rdparty interfaces in OCP please

File: lib/public/http/client/iresponse.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	/**
+	 * @param GuzzleResponse $response
+	 */
+	public function __construct(GuzzleResponse $response);
```
Comment (DeepDiver1975) [inline lib/public/http/client/iresponse.php]: no 3rdparty code in ocp

File: lib/public/http/client/iresponse.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+	/**
+	 * @return \OC\Http\Client\StreamInterface
```
Comment (DeepDiver1975) [inline lib/public/http/client/iresponse.php]: private interface in OCP

File: lib/private/http/client/streaminterface.php (hunk 4 above)
Comment (LukasReschke) [inline lib/private/http/client/streaminterface.php]: So that we don't have to copy the whole StreamInterface code from Guzzle (a lot) and don't have a 3rdparty instance returned to developers? :see_no_evil: 

If you have another suggestion: Open for comments ;-)

Comment (DeepDiver1975) [inline lib/private/http/client/streaminterface.php]: pffff - what about using Guzzle plain to implement the appstore download stuff - which fur sure fixes a true bug. Introducing a public api is something we could do in a later stage.

Makes sense?

Comment (LukasReschke) [inline lib/private/http/client/streaminterface.php]: > Introducing a public api is something we could do in a later stage.

Actually I'd love to use this in other places as well, the stuff we do in our HTTPHelper is sooooooooo wrong on sooooo many levels ;-)

File: lib/public/http/client/iresponse.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+use Guzzle\Http\Message\Response as GuzzleResponse;
+
+/**
+ * Interface IResponse
+ *
+ * @package OC\Http
+ */
+interface IResponse {
+	/**
+	 * @return \OC\Http\Client\StreamInterface
```
Comment (nickvergessen) [inline lib/public/http/client/iresponse.php]: OCP?

File: lib/public/http/client/iresponse.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+namespace OCP\Http\Client;
+
+use Guzzle\Http\Message\Response as GuzzleResponse;
+
+/**
+ * Interface IResponse
+ *
+ * @package OC\Http
```
Comment (nickvergessen) [inline lib/public/http/client/iresponse.php]: OCP

File: lib/private/server.php [hunk 9]
```diff
@@ -646,6 +650,15 @@ function getHTTPHelper() {
 	}
 
 	/**
+	 * Returns an instance of the HTTP client service
+	 *
+	 * @return ClientService
```
Comment (nickvergessen) [inline lib/private/server.php]: type hint should return the OCP interface

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.

**Comment by DeepDiver1975** [comment]
````
00:01:19.649 PHP Warning:  require(/var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/react/promise/src/functions_include.php): failed to open stream: No such file or directory in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/composer/autoload_real.php on line 58
00:01:19.649 PHP Stack trace:
00:01:19.649 PHP   1. {main}() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/occ:0
00:01:19.649 PHP   2. require_once() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/occ:11
00:01:19.649 PHP   3. require_once() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/console.php:14
00:01:19.649 PHP   4. OC::init() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/base.php:1031
00:01:19.649 PHP   5. require_once() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/lib/base.php:473
00:01:19.649 PHP   6. ComposerAutoloaderInit45518c348c80881aa936caecc452b13e::getLoader() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/autoload.php:7
00:01:19.649 PHP   7. composerRequire45518c348c80881aa936caecc452b13e() /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/3rdparty/composer/autoload_real.php:49

**Comment by LukasReschke** [comment]
Yes. Jenkins is too stupid again for that 3rdparty repo. Great fun. Let's create again another PR in 3rdparty for that :angry:

**Comment by LukasReschke** [comment]
Did some rebasing on this PR and the 3rdparty one… Hopefully that is enough for Jenkins.

**Comment by LukasReschke** [comment]
Apparently not……

File: lib/private/http/client/streaminterface.php (hunk 4 above)
Comment (LukasReschke) [inline lib/private/http/client/streaminterface.php]: Will now only return a string. Does not allow all the streaming magic as before but unlikely that we need that too often :see_no_evil: 

(also if you want to have a file as result you can still use `save_to`)

**Comment by LukasReschke** [comment]
I went digging further and found out:

1. ownCloud currently implements proxy handling completely wrong and inconsistent. 
2. The SSL verification implemented by ownCloud is also completely wrong and TOTALLY broken

Thus I extended this PR and added proper proxy handling as well as SSL and certificate handling. Also ownCloud will now ship a list of root certificates to finally get rid of bug reports by those annoying users and enterprise distributions using 10 year old stuff (because the older the more secure!).
Regardless of that: The current approach will append the default root certificates to the users custom `rootcerts.crt`, however, after an update the shipped bundle can change. How can we regenerate this bundle for every user the easiest?  – IMHO a "loop over all directories in `data` and regenerate the certificates if there is a folder `files_external` would be the easiest approach.

Basically this means that if we merge this ownCloud 8.1 will be the first release that properly handles proxies and SSL………

cc @icewind1991 For input
@DeepDiver1975 FYI

**Comment by LukasReschke** [comment]
Regarding the changeset of 4k: No worries, most of these are the added root certificates :see_no_evil:

File: lib/private/http/client/client.php [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+		$this->setDefaultOptions();
+	}
+
+	/**
+	 * Sets the default options to the client
+	 */
+	private function setDefaultOptions() {
+		// Either use default bundle or the user bundle if nothing is specified
+		if($this->certificateManager->listCertificates() !== []) {
+			$dataDir = \OC::$server->getConfig()->getSystemValue('datadirectory');
```
Comment (LukasReschke) [inline lib/private/http/client/client.php]: TODO: Use the injected config

File: apps/files_sharing/lib/external/storage.php [hunk 11]
```diff
@@ -210,37 +210,23 @@ protected function testRemote() {
...
-		$result = curl_exec($ch);
-
-		$status = curl_getinfo($ch, CURLINFO_HTTP_CODE);
-		$errorMessage = curl_error($ch);
-		curl_close($ch);
-		if (!empty($errorMessage)) {
-			throw new \Exception($errorMessage);
-		}
+		// TODO: DI
+		$client = \OC::$server->getHTTPClientService()->newClient();
```
Comment (LukasReschke) [inline apps/files_sharing/lib/external/storage.php]: As commented: use also DI – somehow…

Comment (LukasReschke) [inline apps/files_sharing/lib/external/storage.php]: Also test if it still works :see_no_evil:

**Comment by LukasReschke** [comment]
That said I'd really LOVE to have this in 8.1 before we have to wait yet again 4 months to have proper SSL and proxy support… Pretty annoying and would solve a lot of problems reported by users and customers…

**Comment by LukasReschke** [comment]
For discussion regarding API design:

@MorrisJobke @DeepDiver1975 @PVince81 @th3fallen @Raydiation @icewind1991 @karlitschek @nickvergessen @bantu

**Comment by LukasReschke** [comment]
Will fix the failing tests (basically adjusting the constructor in the tests etc…) once we have agreed to an API.

File: lib/public/http/client/iclient.php (hunk 2 above)
Comment (th3fallen) [inline lib/public/http/client/iclient.php]: Any reason for that @DeepDiver1975 ? Makes no sense to maintain our own copy of that interface?

**Comment by th3fallen** [comment]
:raised_hands: Guzzle ftw

File: lib/public/http/client/iclient.php (hunk 2 above)
Comment (LukasReschke) [inline lib/public/http/client/iclient.php]: The reason there is that we want that applications do not directly use any 3rdparty library cod. The reason is mainly that we had to replace some libraries in the past and replacing a library is pretty hard if you expose the whole function set. If we only expose a limited subset we can easily throw in another replacement without that developers have to adjust anything at all.

Yes – can be a pain – I know. But it's a valid approach that I consider a good solution for us :)

File: lib/private/server.php [hunk 12]
```diff
@@ -651,7 +661,7 @@ function getDb() {
 
 	/**
 	 * Returns an instance of the HTTP helper class
-	 *
+	 * @deprecated Use \OCP\Http\Client\IClientService
```
Comment (nickvergessen) [inline lib/private/server.php]: "Use getHTTPClientService()"

**Comment by nickvergessen** [comment]
Well I guess it makes sense like that

**Comment by nickvergessen** [comment]
Needs rebase

**Comment by LukasReschke** [comment]
@owncloud-bot Retest this please.
